### PR TITLE
Student search backend APIs

### PR DESCRIPTION
**Pull request by Mari-Hayashi**
Added a backend api api/student/search that allows users to search for students based on criteria.

### CODE REVIEW

File: src/server/README.md [hunk 1]
```diff
@@ -728,6 +728,64 @@ Student API
                 -   Unauthorized user -> Status code: 401
                 -   Internal server error -> Status code: 500
 
+    -   api/student/search
+    
+        -   Filter student records based on query and return them.
```
Comment (johan1505) [inline src/server/README.md]: Filters* queries* returns*

File: src/server/README.md [hunk 2]
```diff
@@ -728,6 +728,64 @@ Student API
...
+                            college: {
+                                id: number,
+                                name: string
+                            }
+                        },
+                        courses: {
+                            id: number,
+                            title: string
+                        }[]
+                    } []
```
Comment (johan1505) [inline src/server/README.md]: missing comma at the end

File: src/server/src/modules/student.ts [hunk 3]
```diff
@@ -128,3 +127,52 @@ export const getJobApplications = async (studentId: number) => {
         ])
         .getMany();
 };
+
+export const searchStudents = async (
+    firstName: string,
```
Comment (johan1505) [inline src/server/src/modules/student.ts]: use IStudent['firstName'] type instead of string

File: src/server/src/modules/student.ts [hunk 4]
```diff
@@ -128,3 +127,52 @@ export const getJobApplications = async (studentId: number) => {
...
+    lastName: string,
```
Comment (johan1505) [inline src/server/src/modules/student.ts]: replace types of the arguments of this function to use types from IStudent interface

File: src/server/src/modules/student.ts [hunk 5]
```diff
@@ -128,3 +127,52 @@ export const getJobApplications = async (studentId: number) => {
...
+            firstName: `%${firstName.toLowerCase()}%`,
+        })
+        .andWhere('LOWER(user.lastName) LIKE :lastName', {
+            lastName: `%${lastName.toLowerCase()}%`,
+        })
+        .andWhere('user.email LIKE :email', {
+            email: `%${email}%`,
+        })
+        .andWhere('((NOT :sidExists) OR (:sidExists AND sid LIKE :sid))', {
+            sidExists: sid == "" ? false : true,
```
Comment (johan1505) [inline src/server/src/modules/student.ts]: sidExists: sid !== ""

File: src/server/README.md [hunk 6]
```diff
@@ -728,6 +728,64 @@ Student API
...
+        -   Filter student records based on query and return them.
+        -   Body: None
+        -   Parameters:
+            ```
+            {
+                firstName?: string;
+                lastName?: string;
+                email?: string;
+                sid?: string;
+                departmentIds?: string[];
```
Comment (johan1505) [inline src/server/README.md]: I don't think departmentIds needs to be optional. Getting it as an empty array means the user didn't pick anything for it. so update to   departmentIds: string[];

File: src/server/README.md [hunk 7]
```diff
@@ -728,6 +728,64 @@ Student API
...
+                classStandings?: 'freshman' | 'sophomore' | 'junior' | 'senior'[];
```
Comment (johan1505) [inline src/server/README.md]: Same as departmentIds

File: src/server/src/modules/student.ts [hunk 8]
```diff
@@ -128,3 +127,52 @@ export const getJobApplications = async (studentId: number) => {
...
+            sid: `%${sid}%`,
+        })
+        .andWhere('(NOT :departmentIdsPopulated OR department.id IN (:...departmentIds))', {
+            departmentIdsPopulated: departmentIds[0] != -1,
```
Comment (johan1505) [inline src/server/src/modules/student.ts]: if departmentIds come as an empty array, I think departmentIds[0] will return undefined which is not -1, so you will get a true for your departmentIdsPopulated variable.  Maybe I better check would be departmentIds.length > 0. Also in typescript equality is === and inequality is !==

File: src/server/src/routes/student/index.ts [hunk 9]
```diff
@@ -142,6 +143,68 @@ router.get(
...
+ ******************************************************************************/
+
+interface StudentSearchRequest extends Request {
+    query: {
+        firstName?: string;
+        lastName?: string;
+        email?: string;
+        sid?: string;
+        departmentIds?: string[];
+        classStandings?: classStandings[];
```
Comment (johan1505) [inline src/server/src/routes/student/index.ts]: remove optional

File: src/server/src/routes/student/index.ts [hunk 10]
```diff
@@ -142,6 +143,68 @@ router.get(
...
+        let { firstName, lastName, email, sid, classStandings } = req.query;
+
+        if (!firstName) firstName = "";
+        if (!lastName) lastName = "";
+        if (!email) email = "";
+        if (!sid) sid = "";
+        var departmentIdInts = [];
+        // Pass -1 when the input is empty because it causes a sql parse error
+        // when we pass in an empty array.
+        if (!departmentIds) departmentIdInts = [-1];
```
Comment (johan1505) [inline src/server/src/routes/student/index.ts]: The client will never send departmentIds as undefined, so I don't think we need this. Note: for the client optional for departmentIds means an empty array.

File: src/server/src/routes/student/index.ts [hunk 11]
```diff
@@ -142,6 +143,68 @@ router.get(
...
+    validationMiddleware({ querySchema: studentSearchSchema }),
+    async (req: StudentSearchRequest, res: Response) => {
+        const { departmentIds, page, numOfItems } = req.query;
+        let { firstName, lastName, email, sid, classStandings } = req.query;
+
+        if (!firstName) firstName = "";
+        if (!lastName) lastName = "";
+        if (!email) email = "";
+        if (!sid) sid = "";
+        var departmentIdInts = [];
```
Comment (johan1505) [inline src/server/src/routes/student/index.ts]: var -> const. Note avoid using var, they are bad practice in typescript

File: src/server/src/routes/student/index.ts [hunk 12]
```diff
@@ -142,6 +143,68 @@ router.get(
...
+        // Pass -1 when the input is empty because it causes a sql parse error
+        // when we pass in an empty array.
+        if (!departmentIds) departmentIdInts = [-1];
+        else {
+            for (const id of departmentIds) {
```
Comment (johan1505) [inline src/server/src/routes/student/index.ts]: Nothing wrong here. but using the .map function is a best practice. Like this:

`departmentIdInts = departmentIds.map((id) => parseInt(id, 10));`

I think you also need to pass 10 to tell the parse function you are doing a base 10 conversion.

File: src/server/src/routes/student/index.ts [hunk 13]
```diff
@@ -142,6 +143,68 @@ router.get(
...
+                departmentIdInts.push(parseInt(id));
+            }
+        }
+        if (!classStandings) classStandings = ['Freshman', 'Sophomore', 'Junior', 'Senior'];
```
Comment (johan1505) [inline src/server/src/routes/student/index.ts]: I noticed this array used here maybe it should be declared in entities/students. In that file, please declare and export an array classStandingValues that stores 'Freshman', 'Sophomore', 'Junior', and 'Senior'. Then, import it in this file and use it in your if branch. Also, please update the entity Student class to use that new array (line 36 on entities/student.tsx).

Comment (johan1505) [reply src/server/src/routes/student/index.ts]: Also note that classStandings will never be undefined, so I think your if condition will have to be classStandings.length === 0

**Review by johan1505 — CHANGES_REQUESTED** [review]
Left coments

File: src/server/src/routes/student/index.ts (hunk 10 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: But if the departmentIds array is empty, it causes an SQL parse error. That's why I added -1. What should we do?

File: src/server/src/routes/student/index.ts (hunk 11 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts:175]: But I need to push things into departmentIdInts. Can I push things into const array?

File: src/server/src/routes/student/index.ts (hunk 13 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: How do I update the entity Student class to use the new array? Does this work?
classStandingsStrings = ['Freshman', 'Sophomore', 'Junior', 'Senior'];
classStandings = classStandingsStrings [0] | classStandingsStrings [1] | ...

Comment (johan1505) [reply src/server/src/routes/student/index.ts]:
```
@Column({
        nullable: true,
        type: 'enum',
        enum: classStandingValues, // classStandingValues is the array that contains the class standings values
    })
    classStanding?: classStandings;
```

File: src/server/src/routes/student/index.ts (hunk 10 above)
Comment (johan1505) [reply src/server/src/routes/student/index.ts]: `departmentIdsPopulated: departmentIds[0] != -1`
 if this is changed to 
`departmentIdsPopulated: departmentIds.length !== 0`
do you still get the SQL parse error?

File: src/server/src/modules/student.ts (hunk 8 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/student.ts]: When departmentIds is empty, it causes an sql parse error when I do "department.id IN (:...departmentIds)"
Therefore, when departmentIds is empty, I populate it with -1. So currently, departmentIds is never empty.
Please let me know if you think of another solution.

File: src/server/README.md (hunk 1 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/README.md (hunk 2 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/src/modules/student.ts (hunk 3 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/student.ts]: Done. (Used Student instead of IStudent.)

File: src/server/src/modules/student.ts (hunk 4 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/student.ts]: Done.

File: src/server/src/modules/student.ts (hunk 5 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/student.ts]: Done.

File: src/server/src/routes/student/index.ts (hunk 9 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: Done.

File: src/server/README.md (hunk 6 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/README.md (hunk 7 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/src/modules/student.ts (hunk 8 above)
Comment (Mari-Hayashi) [reply src/server/src/modules/student.ts]: Done.

File: src/server/README.md (hunk 7 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/src/routes/student/index.ts (hunk 10 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: Done.

File: src/server/src/routes/student/index.ts (hunk 12 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: Done.

File: src/server/src/routes/student/index.ts (hunk 13 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: Done.

File: src/server/src/routes/student/index.ts (hunk 11 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/index.ts]: Done.

File: src/server/src/routes/student/schemas.ts [hunk 14]
```diff
@@ -24,12 +24,10 @@ export const studentSearchSchema = Joi.object({
     lastName: Joi.string().allow('').optional(),
     email: Joi.string().allow('').optional(),
     sid: Joi.string().allow('').optional(),
-    departmentIds: Joi.array()
-        .items(Joi.number())
-        .optional(),
-    classStandings: Joi.array()
+    departmentIds: Joi.array().allow('').items(Joi.number()).required(),
```
Comment (johan1505) [inline src/server/src/routes/student/schemas.ts:27]: Question: Why is departmentIds allow to receive an empty string?

**Review by johan1505 — APPROVED** [review]

File: src/server/src/routes/student/schemas.ts (hunk 14 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student/schemas.ts:27]: If I do not have "allow('')", I cannot have an empty array as an input on postman.

**Pull request merged**